File: viewer/IRPanoView.Viewer/Network/ThermalStreamReceiver.cs
using System.Net;
using System.Net.Sockets;
using IRPanoView.Viewer.Network;

namespace IRPanoView.Viewer.Network;

public sealed class ThermalStreamReceiver : IDisposable
{
    public sealed class StreamStats
    {
        public double Fps { get; set; }
        public int LastSequence { get; set; }
        public long Drops { get; set; }
        public string? LastSource { get; set; }
    }

    private readonly object _gate = new();
    private readonly ThermalChunkReassembler _chunkAssembler = new();
    private UdpClient? _client;
    private CancellationTokenSource? _cts;
    private Task? _task;
    private ThermalFrameParser.ParsedFrame? _latest;
    private int _lastSequence = -1;
    private int _frames;
    private long _drops;
    private DateTime _fpsWindowStart = DateTime.UtcNow;
    private double _fps;

    public StreamStats Stats
    {
        get
        {
            lock (_gate)
            {
                return new StreamStats
                {
                    Fps = _fps,
                    LastSequence = _lastSequence,
                    Drops = _drops,
                    LastSource = _latestSource,
                };
            }
        }
    }

    private string? _latestSource;

    public bool IsRunning => _task is { IsCompleted: false };

    public void Start(int port = 8765)
    {
        Stop();
        _cts = new CancellationTokenSource();
        _client = new UdpClient
        {
            Client =
            {
                ExclusiveAddressUse = false,
            },
        };
        _client.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
        _client.Client.Bind(new IPEndPoint(IPAddress.Any, port));
        _task = Task.Run(() => ReceiveLoop(_cts.Token));
    }

    public void Stop()
    {
        _cts?.Cancel();
        try
        {
            _task?.Wait(500);
        }
        catch (AggregateException)
        {
        }

        _client?.Dispose();
        _client = null;
        _cts?.Dispose();
        _cts = null;
        _task = null;
    }

    public bool TryTakeLatest(out ThermalFrameParser.ParsedFrame? frame)
    {
        lock (_gate)
        {
            frame = _latest;
            return frame != null;
        }
    }

    private void ReceiveLoop(CancellationToken token)
    {
        var client = _client ?? throw new InvalidOperationException("UDP client not started");
        while (!token.IsCancellationRequested)
        {
            try
            {
                var remote = new IPEndPoint(IPAddress.Any, 0);
                var packet = client.Receive(ref remote);
                ThermalFrameParser.ParsedFrame? parsed = null;

                if (packet.Length >= ThermalFrameParser.HeaderBytes
                    && BitConverter.ToUInt32(packet.AsSpan(0, 4)) == ThermalFrameParser.Magic)
                {
                    parsed = ThermalFrameParser.Parse(packet);
                }
                else
                {
                    var assembled = _chunkAssembler.Push(packet);
                    if (assembled != null)
                    {
                        parsed = ThermalFrameParser.Parse(assembled);
                    }
                }

                if (parsed == null)
                {
                    continue;
                }

                lock (_gate)
                {
                    if (_lastSequence >= 0 && parsed.Sequence <= _lastSequence)
                    {
                        _drops++;
                    }
                    else if (_lastSequence >= 0 && parsed.Sequence > _lastSequence + 1)
                    {
                        _drops += parsed.Sequence - _lastSequence - 1;
                    }

                    _lastSequence = parsed.Sequence;
                    _latest = parsed;
                    _latestSource = remote.ToString();
                    _frames++;

                    var elapsed = (DateTime.UtcNow - _fpsWindowStart).TotalSeconds;
                    if (elapsed >= 1.0)
                    {
                        _fps = _frames / elapsed;
                        _frames = 0;
                        _fpsWindowStart = DateTime.UtcNow;
                    }
                }
            }
            catch (SocketException) when (token.IsCancellationRequested)
            {
                break;
            }
            catch (ObjectDisposedException)
            {
                break;
            }
        }
    }

    public void Dispose() => Stop();
}

File: src/app/src/main/java/com/vilos/irpanoview/network/ThermalStreamReceiver.kt
package com.vilos.irpanoview.network

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.net.SocketTimeoutException

class ThermalStreamReceiver(
    private val scope: CoroutineScope,
    private val port: Int = PiConnectionManager.UDP_PORT,
) {
    private val _latestFrame = MutableStateFlow<ThermalFrameParser.ParsedFrame?>(null)
    val latestFrame: StateFlow<ThermalFrameParser.ParsedFrame?> = _latestFrame.asStateFlow()

    private val _stats = MutableStateFlow(StreamStats())
    val stats: StateFlow<StreamStats> = _stats.asStateFlow()

    private var job: Job? = null
    private var lastSequence = -1

    data class StreamStats(
        val fps: Double = 0.0,
        val lastSequence: Int = -1,
        val drops: Int = 0,
        val connected: Boolean = false,
    )

    fun start() {
        stop()
        job = scope.launch(Dispatchers.IO) {
            runLoop()
        }
    }

    fun stop() {
        job?.cancel()
        job = null
        _stats.value = StreamStats()
    }

    private suspend fun runLoop() {
        val buffer = ByteArray(512 * 1024)
        var frames = 0
        var drops = 0
        val startMs = System.currentTimeMillis()
        var lastReportMs = startMs

        DatagramSocket(null).use { socket ->
            socket.reuseAddress = true
            socket.soTimeout = 500
            socket.bind(InetSocketAddress(port))
            _stats.value = _stats.value.copy(connected = true)
            while (scope.isActive) {
                try {
                    val packet = DatagramPacket(buffer, buffer.size)
                    socket.receive(packet)
                    val parsed = ThermalFrameParser.parse(
                        buffer.copyOfRange(0, packet.length),
                    ) ?: continue
                    if (parsed.sequence <= lastSequence) {
                        drops++
                        continue
                    }
                    lastSequence = parsed.sequence
                    _latestFrame.value = parsed
                    frames++
                } catch (_: SocketTimeoutException) {
                    // idle
                }

                val now = System.currentTimeMillis()
                if (now - lastReportMs >= 1000L) {
                    val elapsed = (now - lastReportMs).coerceAtLeast(1L)
                    _stats.value = StreamStats(
                        fps = frames * 1000.0 / elapsed,
                        lastSequence = lastSequence,
                        drops = drops,
                        connected = true,
                    )
                    frames = 0
                    lastReportMs = now
                }
            }
        }
        _stats.value = StreamStats()
    }
}
